gremlin-core/src/test/java/org/apache/tinkerpop/gremlin/process/traversal/step/map/ConcatGlobalStepTest.java

```diff
@@ -0,0 +1,49 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.tinkerpop.gremlin.process.traversal.step.map;
+
+import org.apache.tinkerpop.gremlin.process.traversal.Traversal;
+import org.apache.tinkerpop.gremlin.process.traversal.dsl.graph.__;
+import org.apache.tinkerpop.gremlin.process.traversal.step.StepTest;
+import org.junit.Test;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+
+import static org.junit.Assert.assertEquals;
+
+public class ConcatGlobalStepTest extends StepTest {
+
+    @Override
+    protected List<Traversal> getTraversals() {
+        return Collections.singletonList(__.concat("test"));
+    }
+
+    @Test
+    public void testReturnTypes() {
+        assertEquals("whatisthis", __.__("what").concat(__.__("is", "this")).next());
+        assertEquals("thisisatest", __.__("this").concat("is", "a", "test").next());
+        assertEquals("thisisatest", __.__("this", "is", "a", "test").concat().next());
+        // strings in list should be processed and concatenated as well
+        assertEquals("whatisthis?atest", __.__("what", "is", new ArrayList<>(Arrays.asList("this","?"))).concat(__.__("a", "test")).next());
```

Comment:
It's not obvious to me how this behaviour should be defined once List concat is added. Currently if there is an input list of strings we are unfolding it. This makes sense right now but once we have a list concat as well, there is a reasonable argument that we should instead be promoting "what" and "is" into arrays of length 1 and performing a list concat instead.

Reply:
Since list concat is out of scope for this I'm not considering it at the moment. But it would be similar for concat lists in the global context, i.e. all of the items in lists are put into the one list at the end. 
So possibly something like:
```
g.inject([1,2],[3]).concat([4])
==>[1,2,3,4]
```
```
g.inject([1,2],[3]).concat(Scope.local, [4])
==>[1,2,4]
==>[3, 4]
```
I don't believe we'll do promoting of individual items to lists if they aren't in lists, because the definition of list concat require the input to be lists `Returns the concatenation of the incoming array and the traversal or array value passed as a parameter.`. The only dilemma could be if I have string lists, would I potentially only want a string concat as opposed to list concat:
```
g.inject([a,b],[c]).concat([d])
==>[a,b,c,d]
vs
==>abcd
```
But this is more of something to be decided when we do concat list.